server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java

```diff
@@ -145,72 +151,88 @@ private PublishWithJoinResponse handleIncomingPublishRequest(BytesTransportReque
+                );
             } else {
                 final ClusterState lastSeen = lastSeenClusterState.get();
                 if (lastSeen == null) {
                     logger.debug("received diff for but don't have any local cluster state - requesting full state");
                     incompatibleClusterStateDiffReceivedCount.incrementAndGet();
                     throw new IncompatibleClusterStateVersionException("have no local cluster state");
                 } else {
-                    ClusterState incomingState;
-                    try {
-                        final Diff<ClusterState> diff;
-                        final boolean includesLastCommittedData = request.version().onOrAfter(INCLUDES_LAST_COMMITTED_DATA_VERSION);
-                        final boolean clusterUuidCommitted;
-                        final CoordinationMetadata.VotingConfiguration lastCommittedConfiguration;
-
-                        // Close stream early to release resources used by the de-compression as early as possible
-                        try (StreamInput input = in) {
-                            diff = ClusterState.readDiffFrom(input, lastSeen.nodes().getLocalNode());
-                            if (includesLastCommittedData) {
-                                clusterUuidCommitted = in.readBoolean();
-                                lastCommittedConfiguration = new CoordinationMetadata.VotingConfiguration(in);
-                            } else {
-                                clusterUuidCommitted = false;
-                                lastCommittedConfiguration = null;
-                            }
-                            assert input.read() == -1;
-                        }
-                        incomingState = diff.apply(lastSeen); // might throw IncompatibleClusterStateVersionException
-                        if (includesLastCommittedData) {
-                            final var adjustedMetadata = incomingState.metadata()
-                                .withLastCommittedValues(clusterUuidCommitted, lastCommittedConfiguration);
-                            if (adjustedMetadata != incomingState.metadata()) {
-                                incomingState = ClusterState.builder(incomingState).metadata(adjustedMetadata).build();
-                            }
-                        }
-                    } catch (IncompatibleClusterStateVersionException e) {
-                        incompatibleClusterStateDiffReceivedCount.incrementAndGet();
-                        throw e;
-                    } catch (Exception e) {
-                        logger.warn("unexpected error while deserializing an incoming cluster state", e);
-                        assert false : e;
-                        throw e;
-                    }
+                    final ClusterState incomingState = deserializeAndApplyDiff(request, in, lastSeen);
```

Comment:
This is probably controversial, but by factoring out the `deserialiseAndApplyDiff`, incomingState can be made final and used directly in the runBefore lambda. Git also appears to retain the history for the body of `deserialiseAndApply` nicely.

Reply:
Nope seems fine to me, nice idea.